rbx_types/src/attributes/mod.rs

```diff
@@ -0,0 +1,141 @@
+use crate::variant::VariantType;
+use std::convert::TryFrom;
+use thiserror::Error;
+
+mod reader;
+mod writer;
+
+pub use reader::get_attributes;
+pub use writer::attributes_from_map;
+
+macro_rules! create_attribute_type {
+    ({
+        $(
+            $key:ident = $number:tt,
+        )+
+    }) => {
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub(crate) enum AttributeType {
+            $(
+                $key = $number,
+            )+
+        }
+
+        impl TryFrom<VariantType> for AttributeType {
+            type Error = AttributeError;
+
+            fn try_from(variant_type: VariantType) -> Result<Self, Self::Error> {
+                match variant_type {
+                    $(
+                        VariantType::$key => Ok(AttributeType::$key),
+                    )+
+
+                    _ => Err(AttributeError::InvalidVariantType),
+                }
+            }
+        }
+
+        impl TryFrom<u8> for AttributeType {
+            type Error = AttributeError;
+
+            fn try_from(byte: u8) -> Result<Self, Self::Error> {
+                match byte {
+                    $(
+                        $number => Ok(Self::$key),
+                    )+
+
+                    other => Err(AttributeError::InvalidValueType(other))
+                }
+            }
+        }
+    };
+}
+
+create_attribute_type!({
+    String = 0x02,
+    Bool = 0x03,
+    Float32 = 0x05,
+    Float64 = 0x06,
+    UDim = 0x09,
+    UDim2 = 0x0A,
+    BrickColor = 0x0E,
+    Color3 = 0x0F,
+    Vector2 = 0x10,
+    Vector3 = 0x11,
+    NumberSequence = 0x17,
+    ColorSequence = 0x19,
+    NumberRange = 0x1B,
+    Rect = 0x1C,
+});
+
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+pub enum AttributeError {
+    #[error("invalid value type: {0}")]
+    InvalidValueType(u8),
+
+    #[error("invalid brick color: {0}")]
+    InvalidBrickColor(u32),
+
+    #[error("invalid entry key")]
+    InvalidEntryKey,
+
+    #[error("invalid name")]
+    InvalidName,
+
+    #[error("invalid size")]
+    InvalidSize,
+
+    #[error("invalid variant type passed")]
+    InvalidVariantType,
+
+    #[error("no value type was found")]
+    NoValueType,
+
+    #[error("couldn't read bytes to deserialize {0}")]
+    Other(&'static str),
+}
+
+#[cfg(all(feature = "serde", test))]
+mod tests {
+    use super::*;
+
+    // This is taken from rbx-test-files/models/attributes/xml.rbxmx.
+    // This is pasted raw as to not create a circular dependency in test (rbx_types -> rbx_xml/rbx_binary -> rbx_types)
```

Comment:
What do you mean by that? rbx-test-files is stored as a submodule in the root of the workspace, which is how rbx_binary uses it. Is there a reason you can't pull from there (other than the test file not being merged in yet)?

Reply:
The dependency isn't with rbx-test-files, I can include that easily. The problem is with then parsing that data, actually getting the AttributesSerialize *out* of the file. This is what rbx_binary and rbx_xml are for, but then we create a circular dependency (which was my first try, but then things like Variant were mismatched because the ones rbx_binary was giving me were from its own crate).

I *could* potentially just write an XML parser just for this, but it seems a bit unnecessary.

Reply:
Right I didn't think about that. Sorry, carry on!